File: 2021_Samsung_SDS_algorithm_lecture/using_Python/1572.py
"""
input :
10 3
3
4
5
6
7
8
9
10
11
12

output :
60
"""

import sys
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline

PIV = 1 << 18
tree = [0 for _ in range(2 * PIV)]


def query(mid_ind):
    idx = 1
    while idx < PIV:
        idx <<= 1
        if tree[idx] < mid_ind:
            mid_ind -= tree[idx]
            idx += 1
    return idx - PIV


def update(num, chk):
    num += PIV
    tree[num] += chk
    while True:
        num >>= 1
        if num == 0: break

        tree[num] += chk


if __name__ == "__main__":
    N, K = map(int, input().rstrip().split())

    back_up = []
    for i in range(K):
        num = int(input())
        back_up.append(num)
        update(num, 1)

    ans = 0
    try:
        for i in range(K, N + 1):
            mid = query((K + 1) // 2)
            ans += mid

            update(back_up[i - K], -1)
            num = int(input())
            back_up.append(num)
            update(num, 1)
    except:
        pass

    print(ans)

File: 2021_Samsung_SDS_algorithm_lecture/using_Cpp/1572.cpp
/*
input :
10 3
3
4
5
6
7
8
9
10
11
12

output :
60
*/

#include <iostream>
#include <algorithm>
using namespace std;
using ll = long long;

// 전역 변수 선언
#define PIV (1 << 18)
int N, K;
int mid;
ll ans;
int tree[2 * PIV];
int m[2 * PIV]; // sliding window로 이 전 값을 빼내야되기 때문에 정의


int query(int mid){
    int idx = 1;
    while (idx < PIV){
        if (tree[idx *= 2] < mid){
            mid -= tree[idx++];
        }
    }
    return idx - PIV;
}


void update(int num, int chk){
    num += PIV;
    tree[num] += chk;
    while (num >>= 1){
        tree[num] += chk;
    }
    
}

int main(){
    ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    freopen("input.txt", "r", stdin);

    cin >> N >> K;
    for (int i = 0; i < K; i++){
        cin >> m[i];
        update(m[i], 1);
    }

    for (int i = 0; i <= N - K; i++){
        mid = query((K + 1) / 2);
        ans += mid;
        
        update(m[i], -1);
        cin >> m[i + K];
        update(m[i + K], 1);
    }
    cout << ans;

    return 0 ;
}